// @justia/releaser/dist/log/github.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = __importDefault(require("@actions/core"));
class GithubLog {
    constructor() {
        this.console = core_1.default;
    }
    getInput(name, options) {
        return this.console.getInput(name, options);
    }
    setOutput(name, value) {
        return this.console.setOutput(name, value);
    }
    isDebug() {
        return this.console.isDebug();
    }
    debug(message) {
        return this.console.debug(message);
    }
    setFailed(message) {
        return this.console.setFailed(message);
    }
    error(message) {
        return this.console.error(message);
    }
    warning(message) {
        return this.console.warning(message);
    }
    info(message) {
        return this.console.info(message);
    }
    log(message) {
        return this.info(message);
    }
    startGroup(name) {
        return this.console.startGroup(name);
    }
    endGroup() {
        return this.console.endGroup();
    }
    async group(name, fn) {
        return this.console.group(name, fn);
    }
}
exports.default = GithubLog;
