// @justia/releaser/dist/plugins/calver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const release_it_1 = require("release-it");
const calver_1 = __importDefault(require("calver"));
const index_1 = __importDefault(require("../index"));
const log_1 = __importDefault(require("../log"));
const { justia: { calverFormat } } = index_1.default;
class CalverPlugin extends release_it_1.Plugin {
    /**
     * Get calendar versioning format
     */
    getFormat() {
        const { format } = this.getContext();
        return format || calverFormat;
    }
    /**
     * Get latest tag from context
     */
    getLatestTag() {
        return this.getContext('latestTag') || this.config.getContext('latestTag');
    }
    getLatestVersion() {
        const latestTagName = this.getLatestTag();
        const { fallbackVersion } = this.config.options.justia;
        return latestTagName ? latestTagName.replace(/^v/, '') : fallbackVersion;
    }
    getIncrementedVersion({ latestVersion }) {
        let calver;
        try {
            calver = new calver_1.default(this.getFormat(), latestVersion);
        }
        catch (e) {
            return this.getIncrementedVersion({ latestVersion: '' });
        }
        if (calver.get() === latestVersion) {
            calver = calver.inc();
        }
        return calver.get();
    }
    getIncrementedVersionCI({ latestVersion }) {
        return this.getIncrementedVersion({ latestVersion });
    }
    getIncrement({ latestVersion }) {
        return this.getIncrementedVersion({ latestVersion });
    }
    beforeBump() {
        // hotfix: issue with release-it with calver
        // Reference: https://github.com/release-it/release-it/blob/2c66707fcabd2ac360860e14fa607153524f261e/lib/tasks.js#L85
        const versionObj = this.config.getContext('version');
        if (typeof versionObj === 'object' && versionObj.constructor.name === 'SemVer') {
            const parts = this.getFormat().split('.').length;
            const version = versionObj.version.split('.').slice(0, parts).join('.');
            log_1.default.info(`Fixing version...\nCurrent version: ${versionObj.version}\nNew version: ${version}`);
            this.config.setContext({ version: { version } });
        }
    }
}
module.exports = CalverPlugin;
