// @justia/releaser/dist/plugins/github.js
"use strict";
const release_it_1 = require("release-it");
const github_1 = require("@actions/github");
const { GITHUB_TOKEN = '' } = process.env;
class GitHubPlugin extends release_it_1.Plugin {
    constructor() {
        super(...arguments);
        this.octokit = github_1.getOctokit(GITHUB_TOKEN);
        this.isEnabled = false;
        this.owner = '';
        this.repo = '';
    }
    init() {
        const { usePr } = this.options;
        const { baseBranch, headBranch, useVersionBranch } = this.config.options.justia;
        const isSameBranch = headBranch === baseBranch;
        this.isEnabled = usePr && (!isSameBranch || useVersionBranch);
        this.registerPrompts();
    }
    /**
     * Register prompts to CLI version
     */
    registerPrompts() {
        super.registerPrompts({
            create_pull_request: {
                type: 'confirm',
                message: () => 'Do you want to create a Github pull request?',
                default: true
            },
            merge_pull_request: {
                type: 'confirm',
                message: () => 'Do you want to merge the created pull request?',
                default: true
            }
        });
    }
    async afterRelease() {
        if (!this.isEnabled) {
            return Promise.resolve();
        }
        const { version, changelog, repo } = this.config.getContext();
        this.setRepoInformation(repo);
        return this.step({
            enabled: true,
            task: () => this.createPullRequest(version, changelog),
            label: 'Git create pull request',
            prompt: 'create_pull_request'
        })
            .then(({ data }) => {
            const { automergePr } = this.options;
            if (!automergePr) {
                return new Promise(() => null);
            }
            return this.step({
                enabled: true,
                task: () => this.mergePullRequest(data.number),
                label: 'Git merge pull request',
                prompt: 'merge_pull_request'
            });
        })
            .catch((err) => {
            this.log.error(`An error occurred while creating the pull request`);
            this.log.error(err);
            this.debug(err);
        });
    }
    /**
     * Set repo information to local GIT
     */
    setRepoInformation(repo) {
        const { owner, project } = repo;
        this.owner = owner;
        this.repo = project;
    }
    /**
     * Create the Github pull request
     */
    async createPullRequest(version, body) {
        const { baseBranch, headBranch, useVersionBranch } = this.config.options.justia;
        const release = `v${version}`;
        const branch = useVersionBranch ? release : headBranch;
        const base = baseBranch;
        const { owner, repo } = this;
        const pr = {
            title: `Release ${release}`,
            body,
            owner,
            repo,
            head: branch,
            base,
            maintainer_can_modify: true,
            draft: false
        };
        return this.octokit.pulls.create(pr);
    }
    /**
     * Merge the Github pull request
     */
    async mergePullRequest(number) {
        const { owner, repo } = this;
        return this.octokit.pulls.merge({
            owner,
            repo,
            pull_number: number
        });
    }
}
module.exports = GitHubPlugin;
