// @justia/releaser/dist/index.js
"use strict";
/* eslint-disable no-template-curly-in-string */
const RELEASE_VERSION_NAME = 'chore(release): ${version}';
const CHANGELOG_PATTERN = 'git log --pretty=format:"* %s (%h)" ${from}...${to} --no-merges --invert-grep --grep=chore';
/* eslint-enable no-template-curly-in-string */
const CHANGELOG_PRESET = 'angular';
const CHANGELOG_INFILE = 'CHANGELOG.md';
const CALVER_FORMAT = 'YYYY.MM.MICRO';
const TAG_MATCH = '[0-9][0-9][0-9][0-9]*.[0-9]*.[0-9]* --match v[0-9][0-9][0-9][0-9]*.[0-9]*.[0-9]*';
module.exports = {
    git: {
        commitMessage: RELEASE_VERSION_NAME,
        changelog: CHANGELOG_PATTERN,
        requireCommits: true,
        requireUpstream: false,
        tagMatch: TAG_MATCH
    },
    github: {
        release: true
    },
    npm: {
        publish: false
    },
    plugins: {
        '@justia/releaser/plugins/conventional-changelog.js': {
            preset: CHANGELOG_PRESET,
            infile: CHANGELOG_INFILE
        },
        '@justia/releaser/plugins/github.js': {
            usePr: true,
            automergePr: true
        },
        '@justia/releaser/plugins/git-config.js': {
            removeVersionBranch: true
        }
    },
    justia: {
        fallbackVersion: '0.0.0',
        baseBranch: 'master',
        headBranch: 'develop',
        useVersionBranch: true,
        calverFormat: CALVER_FORMAT
    }
};
