// @justia/releaser/utilities.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeDeep = exports.isObject = exports.jsonParseRelaxed = exports.isGitHubAction = exports.dirname = exports.isBoolean = exports.parseInputBoolean = exports.exec = void 0;
const path_1 = __importDefault(require("path"));
const child_process_1 = require("child_process");
const exec = (command) => (0, child_process_1.execSync)(command, { encoding: 'utf8' });
exports.exec = exec;
const parseInputBoolean = (value, defaultValue) => value !== undefined && value !== '' ? value === 'true' : defaultValue;
exports.parseInputBoolean = parseInputBoolean;
const isBoolean = (val) => typeof val === 'boolean';
exports.isBoolean = isBoolean;
exports.dirname = path_1.default.resolve(__dirname);
exports.isGitHubAction = process.env.GITHUB_ACTIONS === 'true';
const jsonParseRelaxed = (badJson) => badJson
    .replace(/:\s*"([^"]*)"/g, (_match, p1) => {
    return `: "${p1.replace(/:/g, '@colon@')}"`;
})
    .replace(/:\s*'([^']*)'/g, (_match, p1) => {
    return `: "${p1.replace(/:/g, '@colon@')}"`;
})
    .replace(/(['"])?([a-z0-9A-Z_]+)(['"])?\s*:/g, '"$2": ')
    .replace(/@colon@/g, ':');
exports.jsonParseRelaxed = jsonParseRelaxed;
const isObject = (item) => {
    return item && typeof item === 'object' && !Array.isArray(item);
};
exports.isObject = isObject;
const mergeDeep = (target, ...sources) => {
    if (!sources.length)
        return target;
    const source = sources.shift();
    if ((0, exports.isObject)(target) && (0, exports.isObject)(source)) {
        Object.keys(source).forEach((key) => {
            if ((0, exports.isObject)(source[key])) {
                if (!target[key])
                    Object.assign(target, { [key]: {} });
                (0, exports.mergeDeep)(target[key], source[key]);
            }
            else if (source[key] !== undefined) {
                Object.assign(target, { [key]: source[key] });
            }
        });
    }
    return (0, exports.mergeDeep)(target, ...sources);
};
exports.mergeDeep = mergeDeep;
